File: bipartite_graph.cpp
// Bipartite using BFS

#include <bits/stdc++.h>
#include <stdio.h>

using namespace std;

bool isBipartite( vector<vector<int>>& graph, int src ){
    int V = graph.size();
    vector<int> color( V, -1 );
    color[src] = 1;

    queue<int> q;
    q.push( src );

    while( !q.empty() ){
        int u = q.front();
        q.pop();

        for( int v : graph[u] ){
            if( color[v] == -1 ){
                color[v] = 1 - color[u];
                q.push( v );
            }else if( color[v] == color[u] ){
                return false;
            }
        }
    }
    return true;
}

int main(){
    vector<vector<int>> graph = {
        { 1, 3 },
        { 0, 2 },
        { 1, 3 },
        { 0, 2 }
    };

    cout << "Is the graph bipartite? " << ( isBipartite( graph, 0 ) ? "Yes" : "No" ) << endl;
    return 0;
}

File: bipartite_graph.java
// Bipartite using BFS

import java.util.*;

public class BipartiteGraph{
    static boolean isBipartite( int[][] graph, int src ){
        int[] color = new int[graph.length];
        Arrays.fill( color, -1 );

        color[src] = 1;
        Queue<Integer> queue = new LinkedList<>();
        queue.add( src );

        while( !queue.isEmpty() ){
            int u = queue.poll();

            for( int v : graph[u] ){
                if( color[v] == -1 ){
                    color[v] = 1 - color[u];
                    queue.add( v );
                }else if( color[v] == color[u] ){
                    return false;
                }
            }
        }
        return true;
    }

    public static void main( String[] args ){
        int[][] graph = {
            { 1, 3 },
            { 0, 2 },
            { 1, 3 },
            { 0, 2 }
        };

        System.out.println( "Is the graph bipartite? " + isBipartite( graph, 0 ) );
    }
}
